File: sdk/python-sdk/verity_sdk/protocols/v1_0/PresentProof.py
from verity_sdk.protocols.Protocol import Protocol
from verity_sdk.utils import COMMUNITY_MSG_QUALIFIER
from verity_sdk.utils.Exeptions import WrongSetupException


class PresentProof(Protocol):
    """
    The PresentProof protocol allows one self-sovereign party ask another self-sovereign party for a private
    and verifiable presentation from credentials they hold. This request can be restricted to certain selectable
    restrictions.
    """
    MSG_FAMILY = 'present-proof'
    """the family name for the message family"""
    MSG_FAMILY_VERSION = '1.0'
    """the version for the message family"""

    PROOF_REQUEST = 'request'
    """Name for 'request' control message"""
    STATUS = 'status'
    """Name for 'status' control message"""
    REJECT = 'reject'
    """Name for 'reject' control message"""
    ACCEPT_PROPOSAL = 'accept-proposal'
    """Name for 'accept-proposal' control message"""
    PRESENTATION_RESULT = 'presentation-result'
    """Name for 'presentation-result' signal message"""
    PROTOCOL_INVITATION = 'protocol-invitation'
    """Name for 'protocol-invitation' signal message"""


    def __init__(self,
                 for_relationship: str,
                 thread_id: str = None,
                 name: str = None,
                 proof_attrs=None,
                 proof_predicates=None,
                 by_invitation: bool = False):
        """
        Args:
            for_relationship (str): the relationship identifier (DID) for the pairwise relationship that will be used
            thread_id (str): the thread id of the already started protocol
            name (str): a human readable name for the given request
            proof_attrs (List[Dict]): an array of attribute based restrictions
            proof_predicates (List[Dict]): an array of predicate based restrictions
            by_invitation (bool): flag to create out-of-band invitation as a part of the PresentProof protocol
        """
        super().__init__(
            self.MSG_FAMILY,
            self.MSG_FAMILY_VERSION,
            msg_qualifier=COMMUNITY_MSG_QUALIFIER,
            thread_id=thread_id
        )
        self.for_relationship = for_relationship
        self.name = name
        self.proof_attrs = proof_attrs
        self.proof_predicates = proof_predicates
        self.by_invitation = by_invitation
        self.created = thread_id is None

    async def request(self, context):
        """
        Directs verity-application to request a presentation of proof.

        Args:
            context (Context): an instance of the Context object initialized to a verity-application agent
        """
        await self.send_message(context, await self.request_msg_packed(context))

    def request_msg(self):
        """
        Creates the control message without packaging and sending it.

        Args:
            context (Context): an instance of the Context object initialized to a verity-application agent

        Return:
            the constructed message (dict object)
        """
        if not self.created:
            raise WrongSetupException('Unable to request presentation when NOT starting the interaction')
        msg = self._get_base_message(self.PROOF_REQUEST)
        self._add_thread(msg)
        self._add_relationship(msg, self.for_relationship)
        msg['name'] = self.name
        msg['proof_attrs'] = self.proof_attrs
        if self.proof_predicates:
            msg['proof_predicates'] = self.proof_predicates
        msg['by_invitation'] = self.by_invitation
        return msg

    async def request_msg_packed(self, context):
        """
        Creates and packages message without sending it.

        Args:
            context (Context): an instance of the Context object initialized to a verity-application agent

        Return:
            the bytes ready for transport
        """
        return await self.get_message_bytes(context, self.request_msg())

    async def status(self, context):
        """
        Ask for status from the verity-application agent

        Args:
            context (Context): an instance of the Context object initialized to a verity-application agent
        """
        await self.send_message(context, await self.status_msg_packed(context))

    def status_msg(self):
        """
        Creates the control message without packaging and sending it.

        Args:
            context (Context): an instance of the Context object initialized to a verity-application agent

        Return:
            the constructed message (dict object)
        """
        msg = self._get_base_message(self.STATUS)
        self._add_thread(msg)
        self._add_relationship(msg, self.for_relationship)
        return msg

    async def status_msg_packed(self, context):
        """
        Creates and packages message without sending it.

        Args:
            context (Context): an instance of the Context object initialized to a verity-application agent

        Return:
            the bytes ready for transport
        """
        return await self.get_message_bytes(context, self.status_msg())

    async def accept_proposal(self, context):
        """
        Directs verity-application to accept the proposed presentation.
        verity application will send presentation request based on the proposal.

        Args:
            context (Context): an instance of the Context object initialized to a verity-application agent
        """
        await self.send_message(context, await self.accept_proposal_msg_packed(context))

    def accept_proposal_msg(self):
        """
        Creates the control message without packaging and sending it.

        Args:
            context (Context): an instance of the Context object initialized to a verity-application agent

        Return:
            the constructed message (dict object)
        """
        msg = self._get_base_message(self.ACCEPT_PROPOSAL)
        self._add_thread(msg)
        self._add_relationship(msg, self.for_relationship)
        return msg

    async def accept_proposal_msg_packed(self, context):
        """
        Creates and packages message without sending it.

        Args:
            context (Context): an instance of the Context object initialized to a verity-application agent

        Return:
            the bytes ready for transport
        """
        return await self.get_message_bytes(context, self.accept_proposal_msg())

    async def reject(self, context, reason):
        """
        Directs verity-application to reject this presentation proof protocol

        Args:
            context (Context): an instance of the Context object initialized to a verity-application agent
            reason: the reason for rejection
        """
        await self.send_message(context, await self.reject_msg_packed(context, reason))

    def reject_msg(self, reason):
        """
        Creates the control message without packaging and sending it.

        Args:
            context (Context): an instance of the Context object initialized to a verity-application agent
            reason: the reason for rejection

        Return:
            the constructed message (dict object)
        """
        msg = self._get_base_message(self.REJECT)
        self._add_thread(msg)
        self._add_relationship(msg, self.for_relationship)
        if reason:
            msg['reason'] = reason
        return msg

    async def reject_msg_packed(self, context, reason):
        """
        Creates and packages message without sending it.

        Args:
            context (Context): an instance of the Context object initialized to a verity-application agent
            reason: the reason for rejection

        Return:
            the bytes ready for transport
        """
        return await self.get_message_bytes(context, self.reject_msg(reason))

File: sdk/nodejs-sdk/src/protocols/v1_0/PresentProof.js
const Protocol = require('../Protocol')
const utils = require('../../utils')
/**
 * An interface for controlling a 1.0 PresentProof protocol.
 * @extends Protocol
 * @see <a href="https://github.com/hyperledger/aries-rfcs/tree/4fae574c03f9f1013db30bf2c0c676b1122f7149/features/0037-present-proof" target="_blank" rel="noopener noreferrer">Aries 0037: Present Proof Protocol 1.0</a>
 */
class PresentProofV1x0 extends Protocol {
  /**
  * Constructor for the 1.0 PresentProof object. This constructor creates an object that is ready to start
  * process of requesting a presentation of proof
  *
  * @param forRelationship the relationship identifier (DID) for the pairwise relationship that will be used
  * @param threadId the thread id of the already started protocol
  * @param name A human readable name for the given request
  * @param attributes An array of attribute based restrictions
  * @param predicates An array of predicates based restrictions (">=", ">", "<=", "<")
  * @param byInvitation flag to create out-of-band invitation as a part of the PresentProof protocol
  * @return 1.0 PresentProof object
  *
  * @property {String} msgFamily - 'present-proof'
  * @property {String} msgFamilyVersion - '1.0'
  * @property {String} msgQualifier - 'Community Qualifier'
  * @property {String} this.msgNames.PROOF_REQUEST - 'request'
  * @property {String} this.msgNames.PRESENTATION_RESULT - 'presentation-result'
  * @property {String} this.msgNames.STATUS - 'status'
  * @property {String} this.msgNames.REJECT - 'reject'
  * @property {String} this.msgNames.PROTOCOL_INVITATION - 'protocol-invitation'
  */
  constructor (forRelationship,
    threadId = null,
    name = '',
    attributes = [],
    predicates = [],
    byInvitation = false
  ) {
    const msgFamily = 'present-proof'
    const msgFamilyVersion = '1.0'
    const msgQualifier = utils.constants.COMMUNITY_MSG_QUALIFIER
    super(msgFamily, msgFamilyVersion, msgQualifier, threadId)

    this.forRelationship = forRelationship
    this.name = name
    this.attributes = attributes
    this.predicates = predicates
    this.byInvitation = byInvitation
    this.created = threadId == null

    this.msgNames.PROOF_REQUEST = 'request'
    this.msgNames.PRESENTATION_RESULT = 'presentation-result'
    this.msgNames.STATUS = 'status'
    this.msgNames.REJECT = 'reject'
    this.msgNames.ACCEPT_PROPOSAL = 'accept-proposal'
    this.msgNames.PROTOCOL_INVITATION = 'protocol-invitation'
  }

  /**
     * Directs verity-application to request a presentation of proof.
     *
     * @param context an instance of the Context object initialized to a verity-application agent
     */
  async request (context) {
    await this.sendMessage(context, await this.requestMsgPacked(context))
  }

  /**
     * Creates and packages message without sending it.
     * @param context an instance of the Context object initialized to a verity-application agent
     * @return the byte array ready for transport
     *
     * @see #request
     */
  async requestMsgPacked (context) {
    return this.getMessageBytes(context, this.requestMsg())
  }

  /**
     * Creates the control message without packaging and sending it.
     * @param context an instance of the Context object initialized to a verity-application agent
     * @return the constructed message (JSON object)
     *
     * @see #request
     */
  requestMsg () {
    if (!this.created) {
      throw new utils.WrongSetupError('Unable to request presentation when NOT starting the interaction')
    }
    var msg = this._getBaseMessage(this.msgNames.PROOF_REQUEST)
    msg['~for_relationship'] = this.forRelationship
    msg.name = this.name
    if (this.attributes) {
      msg.proof_attrs = this.attributes
    }
    if (this.predicates && this.predicates.length > 0) {
      msg.proof_predicates = this.predicates
    }
    msg.by_invitation = this.byInvitation
    msg = this._addThread(msg)
    return msg
  }

  /**
     * Ask for status from the verity-application agent
     *
     * @param context an instance of the Context object initialized to a verity-application agent
     */
  async status (context) {
    await this.sendMessage(context, await this.statusMsgPacked(context))
  }

  /**
     * Creates and packages message without sending it.
     * @param context an instance of the Context object initialized to a verity-application agent
     * @return the byte array ready for transport
     *
     * @see #status
     */
  async statusMsgPacked (context) {
    return this.getMessageBytes(context, this.statusMsg())
  }

  /**
     * Creates the control message without packaging and sending it.
     * @param context an instance of the Context object initialized to a verity-application agent
     * @return the constructed message (JSON object)
     *
     * @see #status
     */
  statusMsg () {
    var msg = this._getBaseMessage(this.msgNames.STATUS)
    msg['~for_relationship'] = this.forRelationship
    msg = this._addThread(msg)
    return msg
  }

  /**
     * Directs verity-application to accept the proposal for presenting the proof.
     * verity-application will send presentation request based on the proposal.
     *
     * @param context an instance of the Context object initialized to a verity-application agent
     * @param reason an human readable reason for the rejection
     */
  async acceptProposal (context) {
    await this.sendMessage(context, await this.acceptProposalMsgPacked(context))
  }

  /**
     * Creates and packages message without sending it.
     * @param context an instance of the Context object initialized to a verity-application agent
     * @return the byte array ready for transport
     *
     * @see #acceptProposal
     */
  async acceptProposalMsgPacked (context) {
    return this.getMessageBytes(context, this.acceptProposalMsg())
  }

  /**
     * Creates the control message without packaging and sending it.
     * @param context an instance of the Context object initialized to a verity-application agent
     * @return the constructed message (JSON object)
     *
     * @see #reject
     */
  acceptProposalMsg () {
    var msg = this._getBaseMessage(this.msgNames.ACCEPT_PROPOSAL)
    msg['~for_relationship'] = this.forRelationship
    msg = this._addThread(msg)
    return msg
  }

  /**
     * Directs verity-application to reject this presentation proof protocol
     *
     * @param context an instance of the Context object initialized to a verity-application agent
     * @param reason an human readable reason for the rejection
     */
  async reject (context, reason) {
    await this.sendMessage(context, await this.rejectMsgPacked(context, reason))
  }

  /**
     * Creates and packages message without sending it.
     * @param context an instance of the Context object initialized to a verity-application agent
     * @return the byte array ready for transport
     *
     * @see #reject
     */
  async rejectMsgPacked (context, reason) {
    return this.getMessageBytes(context, this.rejectMsg(reason))
  }

  /**
     * Creates the control message without packaging and sending it.
     * @param context an instance of the Context object initialized to a verity-application agent
     * @return the constructed message (JSON object)
     *
     * @see #reject
     */
  rejectMsg (reason) {
    var msg = this._getBaseMessage(this.msgNames.REJECT)
    msg['~for_relationship'] = this.forRelationship
    msg = this._addThread(msg)
    if (reason) {
      msg.reason = reason
    }
    return msg
  }
}
module.exports = PresentProofV1x0
